design-proposals/backend-storage-block-migration.md

```diff
@@ -0,0 +1,67 @@
+# Overview
+This proposal is about supporting the migration of all VMs that use a persistent feature, such as persistent vTPM.
+
+To make this document as simple as possible, let's consider that storage in Kubernetes can either be filesystem (FS) or
+block mode, and either ReadWriteOnce (RWO) or ReadWriteMany (RWX) access mode, with storage classes supporting any
+number of combinations of both.
+
+As of KubeVirt 1.3, Here is how each of the 4 combinations is supported by the backend-storage for persistent VM state:
+- RWO FS: supported by backend-storage but makes VMs non-migratable
+- RWX FS: preferred combination, supported by backend-storage and compatible with VM migrations
+- RWO block: not supported yet, but PRs exist to implement it in a non-migratable way 
+- RWX block: same as above, with migratability functional but considered unsafe, because it leads to 2 pods (source and target virt-launchers) mounting the same ext3 filesystem concurrently.
+
+This is the best we can do with the current implementation, which involves mounting the same partition on migration source and target.
```

Comment:
Would you elaborate about partitioning? I see "partitioning complexity" below, but the significance is not clear to me.

What do you mean by "current implementation"? KubeVirt's? libvirt's? Have we considered other (possibly better, yet more intrusive) implementations?

Reply:
swtpm, qemu and/or libvirt don't support saving tpm state on block storage.
libvirt supports saving this state in an in-memory secret but it won't be migrateable.
live migration is only supported with a state saved on shared storage.
There is no plan to support any other implementation. To the best of my knowledge.

To support block storage as backed for vTPM, kubevirt will need to have access to in the swtpm container, build a filesystem on top, and mount it. 
The issues arise when kubevirt attempts to mount this volume on both the source and the target containers. 

I hope it answers your question.


This proposal suggests creating two separate block volumes and allowing qemu to copy the state, however, it will require kubevirt to manage the volumes on migration - which creates other challenges to overcome.

Reply:
To my knowledge, other platforms are fine with a shared filesystem volume for migration or don't support it entirely[1].

IIRC, we asked. @jean-edouard @stu-gott could add your thoughts here?

OpenStack Nova: "Live migration, evacuation, shelving and rescuing of servers with vTPMs is not currently supported."
[1] https://docs.openstack.org/nova/latest/admin/emulated-tpm.html

Reply:
@xpivarc pointed out that it looks like swtpm supports block storage, so I take my previous statement back until we figure out if we can use this option.

Reply:
Indeed, with `--tpmstate backend-uri=file://`, swtpm can serialize its files into something like a block storage device.
This option is however not supported by libvirt. That also doesn't work with EFI.